db/db.go

```diff
@@ -4,40 +4,75 @@ import (
 	"errors"
 )
 
-// ErrNotFound is returned when there is no value associated with a key.
-var ErrNotFound = errors.New("key not found")
+// ErrKeyNotFound is returned when there is no value associated with a key.
+var ErrKeyNotFound = errors.New("key not found")
 
 // ErrEmptyKey is returned when key is empty.
 var ErrEmptyKey = errors.New("key cannot be empty")
 
 // ErrIndexOutOfRange is returned when the iterator index is not in a valid range.
 var ErrIndexOutOfRange = errors.New("iterator index out of range")
 
-// DB for storing key-value tuples. The key must be a string and the value must
-// be a byte slice.
-type DB interface {
-	// Insert a value associated with a key. This will override any existing
-	// value associated with the key.
-	Insert(key string, value []byte) error
+// Codec can do encoding/decoding between arbitrary data object and bytes.
+type Codec interface {
 
-	// Get the value associated with the key. If no value is associated with the
-	// key, then an `ErrNotFound` error will be returned.
-	Get(key string) ([]byte, error)
+	// Encode the object into a slice of bytes.
+	Encode(obj interface{}) ([]byte, error)
 
-	// Delete the value associated with the key.
-	Delete(key string) error
+	// Decode the bytes to its original data object and assign it to the given
+	// variable. Value underlying `value` must be a pointer to the correct
+	// type for object.
+	Decode(data []byte, value interface{}) error
 }
 
-// Iterable is a DB that can iterate over its key-value tuples.
-type Iterable interface {
-	DB
+// Table is a sql-like table for storing key-value pairs. It requires the key
```

Comment:
I would not say "SQL-like" because this is more about "SELECT" and "WHERE" statements. Pretty much all DBs (SQL or not) have tables.